server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -83,33 +78,50 @@ public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateSer
         if (DiscoveryNode.isMasterNode(settings) || DiscoveryNode.isDataNode(settings)) {
             try {
                 ensureNoPre019State();
-                final MetaData metaData = metaStateService.loadFullState();
-                final MetaData upgradedMetaData = upgradeMetaData(metaData, metaDataIndexUpgradeService, metaDataUpgrader);
+                final MetaData metaData = metaStateService.getMetaData();
+                assert metaStateService.hasNoPendingWrites();
+
                 // We finished global state validation and successfully checked all indices for backward compatibility
                 // and found no non-upgradable indices, which means the upgrade can continue.
                 // Now it's safe to overwrite global and index metadata.
-                if (metaData != upgradedMetaData) {
+                if (metaData != null) {
+                    long startNS = System.nanoTime();
+                    final MetaData upgradedMetaData = upgradeMetaData(metaData, metaDataIndexUpgradeService, metaDataUpgrader);
                     if (MetaData.isGlobalStateEquals(metaData, upgradedMetaData) == false) {
                         metaStateService.writeGlobalState("upgrade", upgradedMetaData);
+                    } else {
+                        if (metaStateService.isBwcMode()) {
```

Comment:
`MetaStateService` does not perform metadata upgrade on startup. Actually, it never initiates writes itself. It's the responsibility to `GatewayMetaState` to understand what needs to be written. Here we have pretty sophisticated requirement that we want to write non-upgraded metadata only if we're working in BWC mode. In case we're working in non-BWC mode, we want to re-write manifest only. That's why this flag is exposed.